### PR TITLE
Implement Rop for Pool(mode='max')

### PR DESCRIPTION
**Pull request by AndreasMadsen**
Creating a PR as discussed in #4387

_All credit goes to @nouiz and @rodrigob who implemented this in #3587_

/cc @nouiz

PS: I couldn't get all the test to pass on the master branch without any patches applied.

### CODE REVIEW

**Comment by lamblin** [comment]
I've just had a flash that this Op seems really similar to what DownSampleFactorMaxGradGrad (the gradient through the gradient) does, except this one recomputes the max over its first input, and the other one takes it as an input.
I don't want to delay this PR, just wanted to put it out there, and maybe in the future we can reconcile the two.

File: theano/sandbox/cuda/opt.py [hunk 1]
```diff
@@ -1931,12 +1931,30 @@ def local_gpu_downsample_factor_max_grad(node):
...
             gpu_ds_grad = GpuDownsampleFactorMaxGrad(node.op.ds,
                                                      node.op.ignore_border)
-            return [host_from_gpu(gpu_ds_grad(x.owner.inputs[0],
-                                              as_cuda_ndarray_variable(z),
-                                              as_cuda_ndarray_variable(gz)))]
+            return [host_from_gpu(gpu_ds_grad(x.owner.inputs[0], z, gz))]
 
 
 @register_opt()
+@local_optimizer([downsample.DownsampleFactorMaxRop])
```
Comment (poolio) [inline theano/sandbox/cuda/opt.py]: this should be `pool.DownsampleFactorMaxRop` here and below

**Comment by AndreasMadsen** [comment]
@poolio thanks, this has been fixed. Sorry for the delay, I'm very busy these days.

**Comment by nouiz** [comment]
travis is failing with flake8 errors (easy to fix):

```
/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2377:1: E302 expected 2 blank lines, found 1

/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2405:5: E265 block comment should start with '# '

/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2406:9: E265 block comment should start with '# '

/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2566:1: E302 expected 2 blank lines, found 1

/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1927:5: E129 visually indented line with same indent as next logical line

/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1933:13: E129 visually indented line with same indent as next logical line

/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1940:54: E127 continuation line over-indented for visual indent

/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1943:1: E302 expected 2 blank lines, found 1

/home/travis/build/Theano/Theano/theano/tensor/signal/pool.py:1100:39: E201 whitespace after '['

/home/travis/build/Theano/Theano/theano/tensor/signal/pool.py:1131:5: E303 too many blank lines (2)

/home/travis/build/Theano/Theano/theano/tensor/signal/pool.py:1163:9: E266 too many leading '#' for block comment

/home/travis/build/Theano/Theano/theano/tensor/signal/pool.py:1169:13: E701 multiple statements on one line (colon)

/home/travis/build/Theano/Theano/theano/tensor/signal/pool.py:1172:13: E701 multiple statements on one line (colon)

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:261:26: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:261:29: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:261:32: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:261:35: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:261:38: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:262:25: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:262:27: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:262:29: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:263:25: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:263:27: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:263:29: E231 missing whitespace after ','

/home/travis/build/Theano/Theano/theano/tests/test_rop.py:283:52: E231 missing whitespace after ','
```

After fixing them, you can run the tests to be sure it is well fixed:

nosetests theano/tests/test_flake8.py

thanks

**Comment by AndreasMadsen** [comment]
Travis says there are more issues, however when I run `nosetests theano/tests/test_flake8.py` it reports:
```
ERROR (theano.sandbox.cuda): nvcc compiler not found on $PATH. Check your nvcc installation and try again.
.
----------------------------------------------------------------------
Ran 1 test in 20.176s

OK
```

**Comment by lamblin** [comment]
I'm not sure why it reports that error on your machine, but in any case, here are the issues reported by flake8 on travis:
```
/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2377:1: E302 expected 2 blank lines, found 1
/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2405:5: E265 block comment should start with '# '
/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2406:9: E265 block comment should start with '# '
/home/travis/build/Theano/Theano/theano/sandbox/cuda/blas.py:2566:1: E302 expected 2 blank lines, found 1
/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1927:5: E129 visually indented line with same indent as next logical line
/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1933:13: E129 visually indented line with same indent as next logical line
/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1940:54: E127 continuation line over-indented for visual indent
/home/travis/build/Theano/Theano/theano/sandbox/cuda/opt.py:1943:1: E302 expected 2 blank lines, found 1
```

**Comment by nouiz** [comment]
closing as continued in gh-4781

**Pull request closed without merging**